crates/lovely-core/src/sys.rs

```diff
@@ -1,20 +1,59 @@
-use std::{
-    collections::VecDeque,
-    ffi::{c_void, CString},
-    ptr, slice,
-    sync::LazyLock,
-};
+use std::ptr;
+use std::sync::{LazyLock, OnceLock};
+use std::slice;
+use std::ffi::{c_char, c_void, CString};
+use std::collections::VecDeque;
 
 use itertools::Itertools;
-use libloading::{Library, Symbol};
+use libloading::Library;
 use log::info;
 
-pub const LUA_GLOBALSINDEX: isize = -10002;
+pub static LUA: OnceLock<LuaLib> = OnceLock::new();
+
+pub type LuaState = c_void;
 
+pub const LUA_GLOBALSINDEX: isize = -10002;
 pub const LUA_TNIL: isize = 0;
 pub const LUA_TBOOLEAN: isize = 1;
 
-pub type LuaState = c_void;
+macro_rules! generate {
+    ($libname:ident {
+        $(
+            $vis:vis unsafe extern "C" fn $method:ident($($arg:ident: $ty:ty),*) $(-> $ret:ty)?;
+        )*
+    }) => {
+        pub struct $libname {
+            $(
+                $vis $method: unsafe extern "C" fn($($arg: $ty),*) $(-> $ret)?,
+            )*
+        }
+
+        $(
+            /// # Safety
+            $vis unsafe extern "C" fn $method($($arg: $ty),*) $(-> $ret)? {
+                let lua = LUA.get().unwrap_or_else(|| panic!("Failed to access Lua lib defs"));
+                (lua.$method)($($arg),*)
+            }
+        )*
+    };
+}
+
+generate! (LuaLib {
+    pub unsafe extern "C" fn lua_call(state: *mut LuaState, nargs: isize, nresults: isize);
+    pub unsafe extern "C" fn lua_pcall(state: *mut LuaState, nargs: isize, nresults: isize, errfunc: isize) -> isize;
+    pub unsafe extern "C" fn lua_getfield(state: *mut LuaState, index: isize, k: *const c_char);
+    pub unsafe extern "C" fn lua_setfield(state: *mut LuaState, index: isize, k: *const c_char);
+    pub unsafe extern "C" fn lua_gettop(state: *mut LuaState) -> isize;
+    pub unsafe extern "C" fn lua_settop(state: *mut LuaState, index: isize);
+    pub unsafe extern "C" fn lua_pushvalue(state: *mut LuaState, index: isize);
+    pub unsafe extern "C" fn lua_pushcclosure(state: *mut LuaState, f: *const c_void, n: isize);
+    pub unsafe extern "C" fn lua_tolstring(state: *mut LuaState, index: isize, len: *mut isize) -> *const c_char;
+    pub unsafe extern "C" fn lua_toboolean(state: *mut LuaState, index: isize) -> bool;
+    pub unsafe extern "C" fn lua_topointer(state: *mut LuaState, index: isize) -> *const c_void;
+    pub unsafe extern "C" fn lua_type(state: *mut LuaState, index: isize) -> isize;
+    pub unsafe extern "C" fn lua_typename(state: *mut LuaState, index: isize) -> *const c_char;
+    pub unsafe extern "C" fn lua_isstring(state: *mut LuaState, index: isize) -> bool;
```

Comment:
Since we are reworking this these 5 functions are never called in lovely core, and should be removed.